src/riak_cs_gc_batch.erl

```diff
@@ -0,0 +1,313 @@
+    {KeyListRes, UpdKeyListState} = riak_cs_gc_key_list:next(KeyListState),
+    #gc_key_list_result{bag_id=BagId, batch=Batch} = KeyListRes,
+    _ = lager:debug("Next batch keys: ~p", [Batch]),
+    State?STATE{batch=Batch,
+                key_list_state=UpdKeyListState,
+                bag_id=BagId}.
+
+%% @doc Handle a `batch_complete' event from a GC worker process.
+-spec handle_batch_complete(pid(), #gc_worker_state{}, ?STATE{}) -> ?STATE{}.
+handle_batch_complete(WorkerPid, WorkerState, State) ->
+    ?STATE{
+           worker_pids=WorkerPids,
+           batch_count=BatchCount,
+           batch_skips=BatchSkips,
+           manif_count=ManifestCount,
+           block_count=BlockCount} = State,
+    #gc_worker_state{batch_count=WorkerBatchCount,
+                     batch_skips=WorkerBatchSkips,
+                     manif_count=WorkerManifestCount,
+                     block_count=WorkerBlockCount} = WorkerState,
+    UpdWorkerPids = lists:delete(WorkerPid, WorkerPids),
+    %% @TODO Workout the terminiology for these stats. i.e. Is batch
+    %% count just an increment or represenative of something else.
+    State?STATE{
+                worker_pids=UpdWorkerPids,
+                batch_count=BatchCount + WorkerBatchCount,
+                batch_skips=BatchSkips + WorkerBatchSkips,
+                manif_count=ManifestCount + WorkerManifestCount,
+                block_count=BlockCount + WorkerBlockCount}.
+
+%% @doc Start a GC worker and return the apprpriate next state and
+%% updated state record.
+-spec start_worker(?STATE{}) -> ?STATE{}.
+start_worker(State=?STATE{batch=[NextBatch | RestBatches],
+                          bag_id=BagId,
+                          worker_pids=WorkerPids}) ->
+     case ?GC_WORKER:start_link(BagId, NextBatch) of
+         {ok, Pid} ->
+             State?STATE{batch=RestBatches,
+                         worker_pids=[Pid | WorkerPids]};
+         {error, _Reason} ->
+             State
+     end.
+
+%% @doc Cancel the current batch of files set for garbage collection.
+-spec cancel_batch(?STATE{}) -> any().
+cancel_batch(?STATE{batch_start=BatchStart,
+                    worker_pids=WorkerPids}=_State) ->
+    %% Interrupt the batch of deletes
+    _ = lager:info("Canceled garbage collection batch after ~b seconds.",
+                   [elapsed(BatchStart)]),
+    [riak_cs_gc_worker:stop(P) || P <- WorkerPids].
+
+-spec ok_reply(atom(), ?STATE{}) -> {reply, ok, atom(), ?STATE{}}.
+ok_reply(NextState, NextStateData) ->
+    {reply, ok, NextState, NextStateData}.
+
+try_next_batch(?STATE{batch=Batch} = State) ->
+    State2 = case Batch of
+                 [] ->
+                     maybe_fetch_next_keys(State);
+                 _ ->
+                     State
+             end,
+    case has_batch_finished(State2) of
+        true ->
+            {stop, normal, State2};
+        _ ->
+            %%?debugHere,
+            State3 =  maybe_start_workers(State2),
```

Comment:
Maybe, `maybe_fetch_next_keys` and `maybe_start_workers` should be called interleaved way, even after just single `batch_complete` event. It's because `?STATE.batch` list has at most only one element (under current impl).